### PR TITLE
[GSoC] Added support of Feedback to work with StateSpace.

### PR DESCRIPTION
**Pull request by abhiphile**
Added support of `StateSpace` in `Feedback` connection.
#### Brief description of what is fixed or changed
With this PR two StateSpace systems can be connected in Feedback connection.

#### Release Notes

* physics.control
  * Added support of Feedback to work with StateSpace.

### CODE REVIEW

**Comment by abhiphile** [comment]
I've tested it using `python-control` and it works fine.
# Doc Tests

## SISO DOCS
```python
A1 = np.array([[-1]])  
B1 = np.array([[1]])  
C1 = np.array([[-1]])  
D1 = np.array([1])  
A2 = np.array([[0]])  
B2 = np.array([[1]])  
C2 = np.array([[1]])  
D2 = np.array([[0]])
  
ss1 = control.StateSpace(A1, B1, C1, D1)  
ss2 = control.StateSpace(A2, B2, C2, D2)

>>> print(ss1.feedback(ss2))
<StateSpace>: sys[3]
Inputs (1): ['u[0]']
Outputs (1): ['y[0]']
States (2): ['x[0]', 'x[1]']

A = [[-1. -1.]
     [-1. -1.]]

B = [[1.]
     [1.]]

C = [[-1. -1.]]

D = [[1.]]
```
## MIMO Docs
```python3
A1 = np.array([[4, 1], [2, -3]])  
B1 = np.array([[5, 2], [-3, -3]])  
C1 = np.array([[2, -4], [0, 1]])  
D1 = np.array([[3, 2], [1, -1]])  
A2 = np.array([[-3, 4, 2], [-1, -3, 0], [2, 5, 3]])  
B2 = np.array([[1, 4], [-3, -3], [-2, 1]])  
C2 = np.array([[4, 2, -3], [1, 4, 3]])  
D2 = np.array([[-2, 4], [0, 1]])
ss1 = control.ss(A1, B1, C1, D1)
ss2 = control.ss(A2, B2, C2, D2)
ss1.feedback(ss2)

>>> print(ss1.feedback(ss2))
<StateSpace>: sys[2]
Inputs (2): ['u[0]', 'u[1]']
Outputs (2): ['y[0]', 'y[1]']
States (5): ['x[0]', 'x[1]', 'x[2]', 'x[3]', 'x[4]']

A = [[  3.     -0.75   -3.75  -18.5   -15.   ]
     [  3.5    -4.875   1.125   9.75    9.   ]
     [  3.    -10.25  -11.25  -25.5   -19.   ]
     [ -4.5    16.125   9.125  42.75   36.   ]
     [ -1.5     5.875   3.875  21.25   18.   ]]

B = [[ -0.25    4.75 ]
     [  0.375  -2.625]
     [  0.25    7.25 ]
     [  0.375 -11.625]
     [  0.625  -4.375]]

C = [[  1.     -3.75   -1.75  -10.5    -9.   ]
     [  0.5    -1.625  -1.625  -4.75   -3.   ]]

D = [[-0.25   2.75 ]
     [ 0.125  1.125]]
```
# Unit Tests

## SISO Unit Tests
```python
a1 = np.array([[0, 1], [1, 0]])
b1 = np.array([[0], [1]])
c1 = np.array([[0, 1]])
d1 = np.array([[0]])
a2 = np.array([[1, 0], [0, 1]])
b2 = np.array([[1], [0]])
c2 = np.array([[1, 0]])
d2 = np.array([[1]])
ss1 = control.ss(a1, b1, c1, d1)
ss2 = control.ss(a2, b2, c2, d2)

>>> print(ss1.feedback(ss2))
<StateSpace>: sys[3]
Inputs (1): ['u[0]']
Outputs (1): ['y[0]']
States (4): ['x[0]', 'x[1]', 'x[2]', 'x[3]']

A = [[ 0.  1.  0.  0.]
     [ 1. -1. -1.  0.]
     [ 0.  1.  1.  0.]
     [ 0.  0.  0.  1.]]

B = [[0.]
     [1.]
     [0.]
     [0.]]

C = [[0. 1. 0. 0.]]

D = [[0.]]

>>> print(ss1.feedback(ss2, 1))
<StateSpace>: sys[6]
Inputs (1): ['u[0]']
Outputs (1): ['y[0]']
States (4): ['x[0]', 'x[1]', 'x[2]', 'x[3]']

A = [[0. 1. 0. 0.]
     [1. 1. 1. 0.]
     [0. 1. 1. 0.]
     [0. 0. 0. 1.]]

B = [[0.]
     [1.]
     [0.]
     [0.]]

C = [[0. 1. 0. 0.]]

D = [[0.]]


## With Transfer function
tf1 = control.TransferFunction(np.array([1, 0]), np.array([1, 1]))
>>> print(ss1.feedback(tf1))
<StateSpace>: sys[8]
Inputs (1): ['u[0]']
Outputs (1): ['y[0]']
States (3): ['x[0]', 'x[1]', 'x[2]']

A = [[ 0.  1.  0.]
     [ 1. -1.  1.]
     [ 0.  1. -1.]]

B = [[0.]
     [1.]
     [0.]]

C = [[0. 1. 0.]]

D = [[0.]]
```
## MIMO Unit Tests
```python
a3 = np.array([[4, 1], [2, -3]])  
b3 = np.array([[5, 2], [-3, -3]])  
c3 = np.array([[2, -4], [0, 1]])  
d3 = np.array([[3, 2], [1, -1]])  
a4 = np.array([[-3, 4, 2], [-1, -3, 0], [2, 5, 3]])  
b4 = np.array([[1, 4], [-3, -3], [-2, 1]])  
c4 = np.array([[4, 2, -3], [1, 4, 3]])  
d4 = np.array([[-2, 4], [0, 1]])
ss3 = control.StateSpace(a3, b3, c3, d3)
ss4 = control.StateSpace(a4, b4, c4, d4)
ss3.feedback(ss4)


# Negative Feedback
>>> print(ss3.feedback(ss4))
<StateSpace>: sys[3]
Inputs (2): ['u[0]', 'u[1]']
Outputs (2): ['y[0]', 'y[1]']
States (5): ['x[0]', 'x[1]', 'x[2]', 'x[3]', 'x[4]']

A = [[  3.     -0.75   -3.75  -18.5   -15.   ]
     [  3.5    -4.875   1.125   9.75    9.   ]
     [  3.    -10.25  -11.25  -25.5   -19.   ]
     [ -4.5    16.125   9.125  42.75   36.   ]
     [ -1.5     5.875   3.875  21.25   18.   ]]

B = [[ -0.25    4.75 ]
     [  0.375  -2.625]
     [  0.25    7.25 ]
     [  0.375 -11.625]
     [  0.625  -4.375]]

C = [[  1.     -3.75   -1.75  -10.5    -9.   ]
     [  0.5    -1.625  -1.625  -4.75   -3.   ]]

D = [[-0.25   2.75 ]
     [ 0.125  1.125]]



# Positive Feedback
>>> print(ss3.feedback(ss4, 1))
<StateSpace>: sys[5]
Inputs (2): ['u[0]', 'u[1]']
Outputs (2): ['y[0]', 'y[1]']
States (5): ['x[0]', 'x[1]', 'x[2]', 'x[3]', 'x[4]']

A = [[  0.57142857   8.85714286   1.          -8.          -9.85714286]
     [  4.57142857  -9.64285714  -1.5          3.           5.14285714]
     [ -1.42857143   5.85714286  -4.         -12.         -13.85714286]
     [  1.71428571  -7.92857143  -2.5         18.          24.42857143]
     [  0.28571429  -2.07142857  -0.5         10.          11.57142857]]

B = [[ 0.85714286 -2.42857143]
     [-0.64285714  1.07142857]
     [ 0.85714286 -4.42857143]
     [-1.92857143  6.21428571]
     [-1.07142857  1.78571429]]

C = [[-0.28571429  1.57142857  1.         -4.         -5.57142857]
     [-0.28571429  1.07142857 -0.5        -3.         -2.57142857]]

D = [[ 0.57142857 -1.28571429]
     [ 0.07142857 -0.78571429]]
```

**Comment by abhiphile** [comment]
@Sc0rpi0n101 @faze-geek Please review this PR.

**Comment by abhiphile** [comment]
In this PR I've changed the parent class of `Feedback` from `TransferFunction` to `SISOLinearTimeInvariant`. It was done to work with `StateSpace` properly.

File: sympy/physics/control/lti.py [hunk 1]
```diff
@@ -2739,8 +2784,8 @@ def sensitivity(self):
 
     def doit(self, cancel=False, expand=False, **hints):
         """
-        Returns the resultant transfer function obtained by the
-        feedback interconnection.
+        Returns the resultant transfer function or state space obtained by
+        feedback connection of transfer functions or state space objects.
```
Comment (Sc0rpi0n101) [inline sympy/physics/control/lti.py:2788]: Please add an appropriate example including state space objects too

File: sympy/physics/control/lti.py [hunk 2]
```diff
@@ -2767,6 +2812,33 @@ def doit(self, cancel=False, expand=False, **hints):
...
+            T1 = I_outputs + self.sign * D1 * E_D2
+            T2 = I_inputs + self.sign * E_D2 * D1
+            A = Matrix.vstack(
+            Matrix.hstack(A1 + self.sign * B1 * E_D2 * C1, self.sign * B1 * E_C2),
+            Matrix.hstack(B2 * T1 * C1, A2 + self.sign * B2 * D1 * E_C2)
+            )
+            B = Matrix.vstack(B1 * T2, B2 * D1 * T2)
+            C = Matrix.hstack(T1 * C1, self.sign * D1 * E_C2)
+            D = D1 * T2
+            return StateSpace(A, B, C, D)
```
Comment (Sc0rpi0n101) [inline sympy/physics/control/lti.py:2841]:
```suggestion
            return StateSpace(A, B, C, D)

```

File: sympy/physics/control/lti.py [hunk 3]
```diff
@@ -2577,6 +2585,36 @@ class Feedback(TransferFunction):
     >>> -F2
     Feedback(Series(TransferFunction(-1, 1, s), TransferFunction(2*s**2 + 5*s + 1, s**2 + 2*s + 3, s), TransferFunction(5*s + 10, s + 10, s)), TransferFunction(-1, 1, s), -1)
 
+    Feedback can be used to connect SISO ``StateSpace`` systems together.
```
Comment (Sc0rpi0n101) [inline sympy/physics/control/lti.py:2588]:
```suggestion
    ``Feedback`` can also be used to connect SISO ``StateSpace`` systems.
```

File: sympy/physics/control/lti.py [hunk 4]
```diff
@@ -2903,16 +2977,69 @@ class MIMOFeedback(MIMOLinearTimeInvariant):
     [ -    --- ]
     [ 1    11  ]{t}
 
+    ``MIMOParallel`` can also be used to connect MIMO ``StateSpace`` systems.
```
Comment (Sc0rpi0n101) [inline sympy/physics/control/lti.py:2980]:
```suggestion
    ``MIMOFeedback`` can also be used to connect MIMO ``StateSpace`` systems.
```

File: sympy/physics/control/lti.py [hunk 5]
```diff
@@ -3091,6 +3223,16 @@ def sensitivity(self):
...
             self.sign*_sys1_mat*_sys2_mat).inv()
 
+    @property
+    def num_inputs(self):
+        """Returns the number of inputs."""
+        return self.sys1.num_inputs
+
+    @property
+    def num_outputs(self):
+        """Returns teh number of outputs."""
```
Comment (Sc0rpi0n101) [inline sympy/physics/control/lti.py:3233]:
```suggestion
        """Returns the number of outputs of the system."""
```

File: sympy/physics/control/lti.py [hunk 6]
```diff
@@ -3091,6 +3223,16 @@ def sensitivity(self):
         return (eye(self.sys1.num_inputs) - \
             self.sign*_sys1_mat*_sys2_mat).inv()
 
+    @property
+    def num_inputs(self):
+        """Returns the number of inputs."""
```
Comment (Sc0rpi0n101) [inline sympy/physics/control/lti.py:3228]:
```suggestion
        """Returns the number of inputs of the system."""
```

File: sympy/physics/control/tests/test_lti.py [hunk 7]
```diff
@@ -1008,18 +1008,14 @@ def test_Feedback_functions():
...
     tf1 = TransferFunction(s+1, 1, s)
     tf2 = TransferFunction(s+2, 1, s)
     fd1 = Feedback(tf1, tf2, -1) # Negative Feedback system
     fd2 = Feedback(tf1, tf2, 1) # Positive Feedback system
     unit = TransferFunction(1, 1, s)
 
-    # Checking the type
-    assert isinstance(fd1, TransferFunction)
-    assert isinstance(fd1, Feedback)
-
```
Comment (Sc0rpi0n101) [inline sympy/physics/control/tests/test_lti.py:1022]: You cannot just remove the tests. If you feel the existing tests are not correct anymore. Please add the appropriate type checks that should be correct now.

**Review by Sc0rpi0n101 — APPROVED** [review]

**Pull request merged**